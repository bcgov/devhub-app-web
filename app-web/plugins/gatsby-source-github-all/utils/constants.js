//
// Dev Hub
//
// Copyright Â© 2018 Province of British Columbia
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// Created by Patrick Simonian on 2018-10-12.
//

// source types that map to fetch routines
const SOURCE_TYPES = {
  GITHUB: 'github',
};

const GRAPHQL_NODE_TYPE = 'DevhubSiphon';

const DEFUALT_IGNORES = [
  'node_modules',
  'CONTRIBUTING.md',
  'CODE_OF_CONDUCT.md',
  'LICENSE',
  'CODE OF CONDUCT.md',
  'openshift',
];
// github rest api v3
const GITHUB_API_ENDPOINT = 'https://api.github.com';

const FILETYPES = {
  yml: 'YAML',
  yaml: 'YAML',
  json: 'JSON',
  md: 'Markdown',
  txt: 'text',
  html: 'HTML',
  HTML: 'HTML',
};

const MEDIATYPES = {
  md: 'text/markdown',
  json: 'application/json',
  yaml: 'text/yaml',
  yml: 'text/yaml',
  txt: 'text/html',
  html: 'text/html',
  HTML: 'text/html',
};

const PROCESSABLE_EXTENSIONS = ['.md', '.yml', '.yaml', '.html'];

const MARKDOWN_FRONTMATTER_SCHEMA = {
  title: {
    type: String,
    required: false,
  },
  description: {
    type: String,
    required: true,
  },
  resourcePath: {
    type: String,
    required: false,
  },
  ignore: {
    type: Boolean,
    required: false,
  },
};

const GITHUB_SOURCE_SCHEMA = {
  url: {
    type: String,
    required: true,
  },
  owner: {
    type: String,
    required: true,
  },
  repo: {
    type: String,
    required: true,
  },
};

module.exports = {
  GRAPHQL_NODE_TYPE,
  MARKDOWN_FRONTMATTER_SCHEMA,
  PROCESSABLE_EXTENSIONS,
  DEFUALT_IGNORES,
  GITHUB_API_ENDPOINT,
  GITHUB_SOURCE_SCHEMA,
  FILETYPES,
  MEDIATYPES,
  SOURCE_TYPES,
};
