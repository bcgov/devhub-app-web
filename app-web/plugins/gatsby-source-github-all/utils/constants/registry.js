/*
Copyright 2019 Province of British Columbia

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at 

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

Created by Patrick Simonian
*/

// just a base schema validating root registration properties.
// nothing internal yet.
const REGISTRY_ITEM_SCHEMA = {
  name: {
    type: String,
    required: true,
  },
  sourceProperties: {
    type: Object,
    required: true,
  },
  slug: {
    type: String,
    required: true,
  },
  resourceType: {
    type: String,
    required: false,
  },
  attributes: {
    type: Object,
    required: false,
  },
  description: {
    type: String,
    required: false,
  },
  template: {
    type: String,
    required: false,
  },
  templatePath: {
    type: String,
    required: false,
  },
};

module.exports = {
  REGISTRY_ITEM_SCHEMA,
};
