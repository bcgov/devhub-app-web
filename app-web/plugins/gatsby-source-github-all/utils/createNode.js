/*
Copyright 2019 Province of British Columbia

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at 

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

Created by Patrick Simonian
*/
const { hashString, convertPositionToSortableString } = require('./helpers');
const { GRAPHQL_NODE_TYPE } = require('./constants');

/**
 * creates the object that is passed into the gatsby create node function
 * to create a siphon topic node type
 * @param {Object} topic the topic object
 * @param {String} id the unique id
 */
const createTopicNode = (topic, id, topicContent = null) => {
  const { name, type, description, metadata } = topic;

  const internalContent = topicContent
    ? {
        content: topicContent.content,
        mediaType: topicContent.metadata.mediaType,
      }
    : {};
  return {
    id,
    children: [],
    parent: null,
    name,
    type,
    description: description || '',
    internal: {
      ...internalContent,
      contentDigest: hashString(JSON.stringify(topic)),
      type: GRAPHQL_NODE_TYPE.TOPIC,
    },
    _metadata: {
      position: convertPositionToSortableString(10, metadata.position),
      template: topic.template,
      templateFile: topic.templateFile,
      slug: topic.slug,
      sourceLocations: topic.sourceLocations,
    },
  };
};

/**
 * creates a object that is passed to the gatsby create node function
 * many of these properties are assigned by convention
 * @param {Object} data the siphon node data
 * @param {String} id the unique id
 * @param {String} topicId the topic id that owns this resource
 */
const createSiphonNode = (data, id, topicId) => ({
  id,
  children: [],
  fileName: data.metadata.fileName,
  fileType: data.metadata.fileType,
  devhubId: data.metadata.id,
  name: data.metadata.name,
  owner: data.metadata.owner,
  parent: topicId,
  path: data.path,
  unfurl: data.metadata.unfurl, // normalized unfurled content from various sources https://medium.com/slack-developer-blog/everything-you-ever-wanted-to-know-about-unfurling-but-were-afraid-to-ask-or-how-to-make-your-e64b4bb9254
  topic: {
    name: data.metadata.topic.name, // name of the topic
    type: data.metadata.topic.type,
  },
  source: {
    name: data.metadata.source, // the source-name
    displayName: data.metadata.sourceName, // the pretty name of the 'source'
    sourcePath: data.metadata.sourceURL, // the path to the source
    type: data.metadata.sourceType, // the type of the source
    _properties: data.metadata.sourceProperties, // the source properties mapped to the node from the registry
  },
  resource: {
    path: data.metadata.resourcePath, // either path to a gastby created page based on this node
    type: data.metadata.resourceType, // the base resource type for this see utils/constants.js
    originalSource: data.metadata.originalResourceLocation, // the original location of the resource
  },
  attributes: data.metadata.attributes,
  internal: {
    contentDigest: hashString(JSON.stringify(data)),
    // Optional media type (https://en.wikipedia.org/wiki/Media_type) to indicate
    // to transformer plugins this node has data they can further process.
    mediaType: data.metadata.mediaType,
    // A globally unique node type chosen by the plugin owner.
    type: GRAPHQL_NODE_TYPE.SIPHON,
    // Optional field exposing the raw content for this node
    // that transformer plugins can take and further process.
    content: data.content,
  },
  _metadata: {
    position: convertPositionToSortableString(10, data.metadata.position),
  },
});

module.exports = {
  createSiphonNode,
  createTopicNode,
};
