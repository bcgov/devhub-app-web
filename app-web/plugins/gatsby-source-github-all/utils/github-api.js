//
// Dev Hub
//
// Copyright Â© 2018 Province of British Columbia
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// Created by Patrick Simonian on 2018-10-12.
//
const {
  GITHUB_API_ENDPOINT,
  FILETYPES,
  PROCESSABLE_EXTENSIONS,
  MEDIATYPES,
  DEFUALT_IGNORES,
} = require('./constants');
const chalk = require('chalk'); // eslint-disable-line
const { TypeCheck } = require('@bcgov/common-web-utils'); // eslint-disable-line
const { Base64 } = require('js-base64'); // eslint-disable-line
const fetch = require('node-fetch'); // eslint-disable-line
const ignore = require('ignore'); // eslint-disable-line
/**
 * returns extension of a file name
 * can handle linux type files (which require no extension)
 * @param {String} name 
 * @returns {String} 'readme.md' => md, '.gitignore' => ''
 */
const getExtensionFromName = name =>
  name.slice((Math.max(0, name.lastIndexOf('.')) || Infinity) + 1);
/**
 * 
 * @param {String} name 
 * @returns name without extension
 * .gitignore => .gitignore, readme.md => readme
 */
const getNameWithoutExtension = name => {
  const ext = getExtensionFromName(name);
  if (ext !== '') {
    const re = new RegExp(`.${ext}$`);
    return name.replace(re, '');
  }
  return name;
};
/** returns the name of the file type by its extension
 * if it is an extensionless file, returns ''
 * @param {String} fileName
 * @returns {String} 'the more verbose name'.. 'md' => 'Markdown'
 */

const getNameOfExtensionVerbose = fileName => {
  const ext = getExtensionFromName(fileName);
  return FILETYPES[ext] ? FILETYPES[ext] : '';
};
/**
 * returns media type from extension
 * @param {String} extension 
 */
const getMediaTypeByExtension = extension =>
  MEDIATYPES[extension] ? MEDIATYPES[extension] : '';
/**
 * Using the recursion param, this
 * function attempts to retrieve all directories/files from a repo
 * there is a limit on how deep it can go down a tree
 * as per https://developer.github.com/v3/git/trees/#get-a-tree-recursively
 * @param {String} repo
 * @param {String} owner
 * @param {String} token
 */
const fetchGithubTree = async (repo, owner, token) => {
  try {
    const result = await fetch(
      `${GITHUB_API_ENDPOINT}/repos/${owner}/${repo}/git/trees/master?recursive=1`,
      {
        method: 'GET',
        headers: {
          Authorization: `Bearer ${token}`,
        },
      }
    );
    return await result.json();
  } catch (e) {
    throw e;
  }
};
/**
 * Fetches contents from file
 * note the media type header, it converts what would have been a
 * b64 encoded string of the file contents into either raw string data or json
 * @param {String} repo 
 * @param {String} owner 
 * @param {String} path 
 * @param {String} token 
 */
const fetchFile = async (repo, owner, path, token) => {
  try {
    const result = await fetch(
      `${GITHUB_API_ENDPOINT}/repos/${owner}/${repo}/contents/${path}`,
      {
        method: 'GET',
        headers: {
          Authorization: `Bearer ${token}`,
          'X-GitHub-Media-Type': 'Accept: application/vnd.github.v3.raw+json',
        },
      }
    );
    return await result.json();
  } catch (e) {
    throw e;
  }
};
/**
 * fetches .devhubignore file from a repo
 * @param { String } repo 
 * @param { String } owner 
 * @param { String } token 
 */
const fetchIgnoreFile = async (repo, owner, token) => {
  const ignoreFile = await fetchFile(repo, owner, '/.devhubignore', token);
  return ignoreFile.content
    ? Base64.decode(ignoreFile.content).split('\n')
    : [];
};
/**
   * filters an array of github graphql entries by their extensions
   * the filtering compares the object.name property with a regex test
   * @param {Array} entries 
   * @param {Array} extensions (defaults to [.md])
   */
const filterFilesByExtensions = (entries, extensions = ['.md']) => {
  // ensure entries is an array of objects
  if (!TypeCheck.isArray(entries) || !entries.every(TypeCheck.isObject)) {
    throw new Error('entries are invalid');
  }

  if (!TypeCheck.isArray(extensions) || !extensions.every(TypeCheck.isString)) {
    throw new Error('extensions must be an array of strings');
  }
  // ensure extensions are of correct pattern
  if (!extensions.every(ext => /\.\w+$/.test(ext))) {
    throw new Error('extensions must have shape /\\.w+$/');
  }
  // convert extensions into a regex expression
  const re = new RegExp(`(${extensions.join('|')})$`);
  // would have shape like /(.md|.txt)$/
  // filter entries
  return entries.filter(entry => re.test(entry.path));
};
/**
   * filters out directories from array of github graph ql entries
   * directories have type of 'tree'
   * @param {Array} entries 
   */
const filterFilesFromDirectories = entries => {
  // ensure entries is an array of objects
  if (!TypeCheck.isArray(entries) || !entries.every(TypeCheck.isObject)) {
    throw new Error('entries are invalid');
  }
  // only tree type entries that don't start with a . (ignore hidden folders)
  return entries.filter(entry => entry.type === 'blob');
};

/**
 * returns a flattened array of all files in a repository
 * by recursive graphql query to each directory in repo (accomplished via breadth first-search)
 * @param {String} repo 
 * @param {String} owner 
 * @param {String} name // name of repo as per registry 
 * @param {String} token 
 */
// eslint-disable-next-line
const getFilesFromRepo = async (repo, owner, name, token) => {
  try {
    // ignore filtering
    const ig = ignore().add(DEFUALT_IGNORES);
    // create graphql string for finding all files in a directory
    const data = await fetchGithubTree(repo, owner, token);
    // filter out files by extensions
    let filesToFetch = filterFilesFromDirectories(data.tree);
    // filter out files that aren't markdown
    filesToFetch = filterFilesByExtensions(
      filesToFetch,
      PROCESSABLE_EXTENSIONS
    );
    // fetch ignore file if exists
    const repoIgnores = await fetchIgnoreFile(repo, owner, token);
    ig.add(repoIgnores);
    // filter out files that are apart of ignore
    filesToFetch = filesToFetch.filter(file => !ig.ignores(file.path));
    // retrieve contents for each file
    const filesWithContents = filesToFetch.map(file =>
      fetchFile(repo, owner, file.path, token)
    );
    const filesResponse = await Promise.all(filesWithContents);
    // for some reason the accept header is not returning with raw content so we will decode
    // the default base 64 encoded content
    // also adding some additional params
    const files = filesResponse.map(f => {
      const ext = getExtensionFromName(f.name);
      return {
        ...f,
        content: Base64.decode(f.content),
        metadata: {
          sourceName: name,
          source: repo,
          owner,
          name: getNameWithoutExtension(f.name),
          fileType: getNameOfExtensionVerbose(f.name),
          fileName: f.name,
          mediaType: getMediaTypeByExtension(ext),
          extension: ext,
        },
      };
    });
    return files;
  } catch (e) {
    // eslint-disable-next-line
    console.error(chalk`
      {red.bold ERROR!! in Gatsby Source Github All} 

      unable to fetch files from repo ${repo}.

      Perhaps you should check spelling of repo parameters..

      {green.underline repo}: ${repo}
      {green.underline owner}: ${owner}

      {yellow if this doesn't resolve the issue either the api token is invalid
      or the build is failing to connect to the github api}
    `);
    throw e;
  }
};

module.exports = {
  getFilesFromRepo,
  getExtensionFromName,
  getNameWithoutExtension,
  getNameOfExtensionVerbose,
  fetchGithubTree,
  fetchFile,
  filterFilesFromDirectories,
  filterFilesByExtensions,
};
