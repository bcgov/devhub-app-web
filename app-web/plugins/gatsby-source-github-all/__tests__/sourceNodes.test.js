//
// Dev Hub
//
// Copyright Â© 2018 Province of British Columbia
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// Created by Patrick Simonian on 2018-10-12.
//
import {
  checkRegistry,
  getRegistry,
  filterIgnoredResources,
  sourcesAreValid,
  mapInheritedSourceAttributes,
  getFetchQueue,
  normalizePersonas,
  processSource,
  getContentForTopic,
  processTopic,
} from '../sourceNodes';
import { createSiphonNode, createTopicNode } from '../utils/createNode';
import { GRAPHQL_NODE_TYPE, TOPIC_TYPES, TOPIC_TEMPLATES } from '../utils/constants';
import {
  GRAPHQL_NODES_WITH_REGISTRY,
  GRAPHQL_NODES_WITHOUT_REGISTRY,
  REGISTRY,
  REGISTRY_WITH_TOPIC,
  TOPIC_OBJ_FROM_FETCH_QUEUE,
  WEB_SOURCE,
  PROCESSED_WEB_SOURCE,
  PROCESSED_FILE_MD,
  SOURCE_REGISTRY_TYPE,
} from '../__fixtures__/fixtures';
import { validateSourceRegistry, fetchFromSource } from '../utils/fetchSource';
import {
  isSourceTopic,
  hashString,
  validateRegistryItemAgainstSchema,
  validateAgainstSchema,
  newTopic,
  assignPositionToTopic,
  assignPositionToSource,
  convertPositionToSortableString,
  assignPositionToResource,
} from '../utils/helpers';

jest.mock('../utils/helpers');
jest.mock('crypto');
jest.mock('../utils/fetchSource.js');
fetchFromSource.mockReturnValue(Promise.resolve([PROCESSED_WEB_SOURCE]));
newTopic.mockImplementation((topic, props) => ({ ...topic, ...props }));
assignPositionToTopic.mockImplementation(topic => () => ({
  metadata: { position: [0] },
}));

assignPositionToResource.mockImplementation(source => resource => ({
  ...resource,
  metadata: { ...resource.metadata, position: [0] },
}));

convertPositionToSortableString.mockImplementation((padding, position) => position.join('.'));

assignPositionToSource.mockImplementation(source => () => ({ metadata: { position: [0, 0] } }));

describe('gatsby source github all plugin', () => {
  afterEach(() => {
    isSourceTopic.mockReset();
    fetchFromSource.mockClear();
  });

  test('filterIgnoreResources returns a filtered set of sources', () => {
    const sources = [
      {
        metadata: {
          ignore: true,
        },
      },
      {
        metadata: {
          ignore: false,
        },
      },
      {
        metadata: {
          apple: true,
        },
      },
    ];

    const expected = [
      {
        metadata: {
          ignore: false,
        },
      },
      {
        metadata: {
          apple: true,
        },
      },
    ];

    expect(filterIgnoredResources(sources)).toEqual(expected);
  });

  test('getRegistry returns the registry', () => {
    const getNodes = jest.fn(() => GRAPHQL_NODES_WITH_REGISTRY);
    const registry = getRegistry(getNodes, SOURCE_REGISTRY_TYPE);
    expect(registry).toBeDefined();
    expect(registry.length).toBeGreaterThan(0);
  });

  test('getRegistry throws if no registry exists', () => {
    const getNodes = jest.fn(() => GRAPHQL_NODES_WITHOUT_REGISTRY);
    expect(() => getRegistry(getNodes, SOURCE_REGISTRY_TYPE)).toThrow('Registry not found');
  });

  test('getRegistry returns with apecified registry json file', () => {
    const getNodes = jest.fn(() => GRAPHQL_NODES_WITH_REGISTRY);
    const registry = getRegistry(getNodes, SOURCE_REGISTRY_TYPE);
    expect(registry).toBeDefined();
  });

  test('checkRegistry returns true if sources are valid', () => {
    validateSourceRegistry.mockReturnValue(true);
    validateRegistryItemAgainstSchema.mockReturnValue(true);
    expect(checkRegistry(REGISTRY)).toBe(true);
  });

  test("checkRegistry throws if sources don't exist or if sources are invalid", () => {
    const BAD_REGISTRY = null;

    expect(() => checkRegistry(BAD_REGISTRY)).toThrow(
      'Error in Gatsby Source Github All: registry is not valid. One or more repos may be missing required parameters',
    );

    validateSourceRegistry.mockReturnValue(false);

    expect(() => checkRegistry(REGISTRY)).toThrow(
      'Error in Gatsby Source Github All: registry is not valid. One or more repos may be missing required parameters',
    );
  });

  test('createSiphonNode returns data', () => {
    const file = {
      metadata: {
        position: [1, 1, 1],
        topic: {
          name: 'foo',
          type: TOPIC_TYPES.CURATED,
        },
        id: 'file-stub-id',
        name: 'test',
        source: 'something/something',
        sourceName: 'something',
        owner: 'Billy Bob',
        fileType: 'Markdown',
        fileName: 'test.md',
        mediaType: 'application/test',
        extension: '.md',
        attributes: {
          labels: 'component',
          personas: ['Developer'],
          tags: ['foo'],
        },
        originalResourceLocation:
          'https://github.com/awesomeOrg/awesomeRepo/blob/master/public/manifest.json',
        sourceProperties: {
          branch: 'master',
          repo: 'foo',
          owner: 'bar',
        },
      },
      name: 'test',
      path: '/test.md',
      sha: '123',
      size: 317,
      url:
        'https://api.github.com/repos/awesomeOrg/awesomeRepo/contents/public/manifest.json?ref=master',
      html_url: 'https://github.com/awesomeOrg/awesomeRepo/blob/master/public/manifest.json',
      git_url: 'https://api.github.com/repos/awesomeOrg/awesomeRepo/git/blobs/123',
      download_url:
        'https://raw.githubusercontent.com/awesomeOrg/awesomeRepo/master/public/manifest.json',
      type: 'file',
      content: 'content',
      _links: {
        self:
          'https://api.github.com/repos/awesomeOrg/awesomeRepo/contents/public/manifest.json?ref=master',
        git: 'https://api.github.com/repos/awesomeOrg/awesomeRepo/git/blobs/123',
        html: 'https://github.com/awesomeOrg/awesomeRepo/blob/master/public/manifest.json',
      },
    };

    hashString.mockReturnValue(null);

    expect(createSiphonNode(file, '123', 'foo')).toMatchSnapshot();
  });

  test('createTopicNode returns an object', () => {
    const result = createTopicNode(TOPIC_OBJ_FROM_FETCH_QUEUE, '123');
    const expected = {
      id: '123',
      name: TOPIC_OBJ_FROM_FETCH_QUEUE.name,
      type: TOPIC_OBJ_FROM_FETCH_QUEUE.type,
      description: TOPIC_OBJ_FROM_FETCH_QUEUE.description,
      children: [],
      parent: null,
      internal: {
        contentDigest: null, // hash string called here
        type: GRAPHQL_NODE_TYPE.TOPIC,
      },
      _metadata: {
        position: TOPIC_OBJ_FROM_FETCH_QUEUE.metadata.position.join('.'),
        slug: TOPIC_OBJ_FROM_FETCH_QUEUE.slug,
        template: TOPIC_TEMPLATES.DEFAULT,
        templateFile: undefined,
        sourceLocations: undefined,
      },
    };
    hashString.mockReturnValue(null);
    expect(result).toEqual(expected);
  });

  test('createTopicNode returns an internal mime type of markdown plus content is topic content is passed in', () => {
    const result = createTopicNode(TOPIC_OBJ_FROM_FETCH_QUEUE, '123', {
      content: 'hello world',
      metadata: {
        mediaType: 'text/markdown',
      },
    });

    const expected = {
      id: '123',
      name: TOPIC_OBJ_FROM_FETCH_QUEUE.name,
      type: TOPIC_OBJ_FROM_FETCH_QUEUE.type,
      description: TOPIC_OBJ_FROM_FETCH_QUEUE.description,
      children: [],
      parent: null,
      internal: {
        contentDigest: null, // hash string called here
        type: GRAPHQL_NODE_TYPE.TOPIC,
        content: 'hello world',
        mediaType: 'text/markdown',
      },
      _metadata: {
        position: TOPIC_OBJ_FROM_FETCH_QUEUE.metadata.position.join('.'),
        slug: TOPIC_OBJ_FROM_FETCH_QUEUE.slug,
        template: TOPIC_TEMPLATES.DEFAULT,
        templateFile: undefined,
        sourceLocations: undefined,
      },
    };
    hashString.mockReturnValue(null);
    expect(result).toEqual(expected);
  });

  test('sourcesAreValid handles topics of sources', () => {
    validateSourceRegistry.mockClear();
    validateSourceRegistry.mockReturnValue(true);
    isSourceTopic.mockReturnValueOnce(true);

    sourcesAreValid(REGISTRY_WITH_TOPIC);
    expect(validateSourceRegistry).toHaveBeenCalledTimes(
      REGISTRY_WITH_TOPIC[0].sourceProperties.sources.length,
    );
  });

  test('maps inherited source props', () => {
    const rootSource = {
      name: 'foo',
      resourceType: 'Designer',
    };
    const childSource = {
      sourceType: 'github',
      sourceProperties: {
        repo: 'blah',
        owner: 'foo',
        url: '/',
      },
    };

    const result = mapInheritedSourceAttributes(rootSource, childSource);
    expect(result.name).toBe(rootSource.name);
    expect(result.resourceType).toBe(rootSource.resourceType);
    expect(result.attributes).toEqual({
      personas: [],
    });
  });

  // get fetch queue should return a list of topics that contain a list of sources to fetch
  test('creates a fetch queue with for a source topic', async () => {
    isSourceTopic.mockReturnValue(false);
    const result = await getFetchQueue(REGISTRY);
    expect(result.length).toBe(REGISTRY.length);
    expect(result[0].sources.length).toBe(1);
  });

  test('creates a fetch queue with for a curated topic', async () => {
    isSourceTopic.mockReturnValue(true);
    const result2 = await getFetchQueue(REGISTRY_WITH_TOPIC);
    expect(result2.length).toBe(REGISTRY_WITH_TOPIC.length);

    expect(result2[0].sources.length).toBe(REGISTRY_WITH_TOPIC[0].sourceProperties.sources.length);
  });

  test('getFetchQueue passes the correct topic type', async () => {
    isSourceTopic.mockReturnValue(false);
    const result = await getFetchQueue(REGISTRY);
    expect(result[0].type).toBe(TOPIC_TYPES.github);

    isSourceTopic.mockReturnValue(true);
    const result2 = await getFetchQueue(REGISTRY_WITH_TOPIC);
    expect(result2[0].type).toBe(TOPIC_TYPES.CURATED);
  });

  test('normalize personas converts persona into personas when alone', () => {
    const attributes = {
      persona: 'Designer',
    };

    const expected = {
      persona: 'Designer',
      personas: ['Designer'],
    };

    expect(normalizePersonas(attributes)).toEqual(expected);
  });

  test('normalize personas leaves personas unchanged when valid', () => {
    const attributes = {
      personas: ['Designer'],
    };

    const expected = {
      personas: ['Designer'],
    };

    expect(normalizePersonas(attributes)).toEqual(expected);
  });

  test('normalize personas uses personas over persona when there is conflict', () => {
    const attributes = {
      personas: ['Designer'],
      persona: 'Developer',
    };

    const expected = {
      personas: ['Designer'],
      persona: 'Developer',
    };

    expect(normalizePersonas(attributes)).toEqual(expected);
  });

  test('normalize personas returns personas as empty array when invalid', () => {
    const attributes = {
      personas: 12,
      persona: 123,
    };

    const expected = {
      personas: [],
      persona: 123,
    };

    expect(normalizePersonas(attributes)).toEqual(expected);
  });

  test('returns a list of source node objects', async () => {
    const createNodeId = jest.fn(() => 1);
    const createNode = jest.fn(node => node);
    const result = await processSource(WEB_SOURCE, createNodeId, createNode, {});

    expect(result instanceof Array).toBe(true);
    expect(result.length).toBe(1);

    const node = result[0];

    expect(typeof node).toBe('object');
    // validate the 'node' has properties that would only exist if the createNode fn was called to create the object
    expect(node.id).toBeDefined();
  });

  // eslint-disable-next-line no-console
  console.log('skipping SourceNodes.topicObject test, this feature is deprecated');
  test.skip('returns a topic object', async () => {
    const createNodeId = jest.fn(() => 1);
    const createNode = jest.fn(node => node);
    const createParentChildLink = jest.fn();

    const result = await processTopic(
      TOPIC_OBJ_FROM_FETCH_QUEUE,
      createNodeId,
      createNode,
      createParentChildLink,
      {},
    );

    expect(typeof result).toBe('object');
    // assert we get the original topic properties back
    expect(result.name).toBe(TOPIC_OBJ_FROM_FETCH_QUEUE.name);
    // assert that its actually a grpahql node by again checking if an id was added
    expect(result.id).toBeDefined();
  });

  test('calls the fetch source routine for each source in a topic object', async () => {
    const createNodeId = jest.fn(() => 1);
    const createNode = jest.fn(node => node);
    const createParentChildLink = jest.fn();

    await processTopic(
      TOPIC_OBJ_FROM_FETCH_QUEUE,
      createNodeId,
      createNode,
      createParentChildLink,
      {},
    );
    // it should be called for as many sources that exist within the topic
    expect(fetchFromSource).toHaveBeenCalledTimes(TOPIC_OBJ_FROM_FETCH_QUEUE.sources.length);
  });

  test('establishes a parent child link for each resource fetched for a topic', async () => {
    const createNodeId = jest.fn(() => 1);
    const createNode = jest.fn(node => node);
    const createParentChildLink = jest.fn();

    await processTopic(
      TOPIC_OBJ_FROM_FETCH_QUEUE,
      createNodeId,
      createNode,
      createParentChildLink,
      {},
    );

    // fetch from source returns a single web source node
    // in total only 1 resource was returned from all sources fetched for the fixtured topic
    // therefor the createparent child link should only be called once
    // expect(createParentChildLink).toHaveBeenCalledTimes(1);
    // TOPICS NO LONGER CREATED BY THIS PLUGIN !
  });

  test('getContentForTopic returns data', async () => {
    fetchFromSource.mockReturnValueOnce(Promise.resolve([PROCESSED_FILE_MD]));
    validateAgainstSchema.mockReturnValueOnce({
      isValid: true,
      messages: [],
    });

    const topicSource = {
      repo: 'foo',
      owner: 'bar',
      file: 'file.md',
    };

    const data = await getContentForTopic(topicSource, { token: 123 }, 'topic name');
    expect(data).toEqual(PROCESSED_FILE_MD);
  });

  test('getContentForTopic returns {} when invalid source properties', async () => {
    fetchFromSource.mockReturnValueOnce(Promise.resolve([PROCESSED_FILE_MD]));
    validateAgainstSchema.mockReturnValueOnce({
      isValid: false,
      messages: [],
    });

    const topicSource = {
      repo: 'foo',
      owner: 'bar',
      file: 'file.md',
    };

    const data = await getContentForTopic(topicSource, { token: 123 }, 'topic name');
    expect(data).toEqual({});
  });
});
