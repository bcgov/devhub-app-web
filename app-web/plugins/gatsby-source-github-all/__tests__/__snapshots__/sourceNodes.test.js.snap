// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`gatsby source github all plugin createSiphonNode returns data 1`] = `
Object {
  "_metadata": Object {
    "position": "1.1.1",
  },
  "attributes": Object {
    "labels": "component",
    "personas": Array [
      "Developer",
    ],
    "tags": Array [
      "foo",
    ],
  },
  "children": Array [],
  "devhubId": "file-stub-id",
  "fileName": "test.md",
  "fileType": "Markdown",
  "id": "123",
  "internal": Object {
    "content": "content",
    "contentDigest": null,
    "mediaType": "application/test",
    "type": "DevhubSiphon",
  },
  "name": "test",
  "owner": "Billy Bob",
  "parent": "foo",
  "path": "/test.md",
  "resource": Object {
    "originalSource": "https://github.com/awesomeOrg/awesomeRepo/blob/master/public/manifest.json",
    "path": undefined,
    "type": undefined,
  },
  "source": Object {
    "_properties": Object {
      "branch": "master",
      "owner": "bar",
      "repo": "foo",
    },
    "displayName": "something",
    "name": "something/something",
    "sourcePath": undefined,
    "type": undefined,
  },
  "topic": Object {
    "name": "foo",
    "type": "curated",
  },
  "unfurl": undefined,
}
`;
